ipv6/prefix.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// Prefix implements PrefixI
+func (me Prefix) Prefix() Prefix {
+	return me
+}
+
+// ToNetIPNet returns a *net.IPNet representation of this prefix
+func (me Prefix) ToNetIPNet() *net.IPNet {
+	return &net.IPNet{
+		IP:   me.addr.ToNetIP(),
+		Mask: net.CIDRMask(me.Length(), addressSize),
+	}
+}
+
+// lessThan reports whether this Prefix comes strictly before `other`
+// lexigraphically.
+func (me Prefix) lessThan(other Prefix) bool {
+	meNet, otherNet := me.Network(), other.Network()
+	if meNet.addr != otherNet.addr {
+		return meNet.addr.lessThan(otherNet.addr)
+	}
+	if me.length == other.length {
+		return me.Host().addr.lessThan(other.Host().addr)
+	}
+	return me.length < other.length
+}
+
+// Length returns the number of leading 1s in the mask.
+func (me Prefix) Length() int {
+	return int(me.length)
+}
+
+// Mask returns a new Address with 1s in the first `length` bits and then 0s
+// representing the network mask for this prefix.
+func (me Prefix) Mask() Mask {
+	return lengthToMask(me.Length())
+}
+
+// Network returns a new Prefix with all bits after `length` zeroed out so that
+// only the bits in the `network` part of the prefix are present. Note that
+// this method ignores special cases where a network address doesn't make
+// sense like in a host route or point-to-point prefix (/128 and /127). It just
+// does the math.
+func (me Prefix) Network() Prefix {
+	network := me.addr.ui.and(me.Mask().ui)
+	return Prefix{
+		addr: Address{
+			ui: network,
+		},
+		length: me.length,
+	}
+}
+
+// Broadcast returns a new Prefix with all bits after `length` set to 1s. Note
+// that this method ignores special cases where a broadcast address doesn't
+// make sense like in a host route or point-to-point prefix (/128 and /127). It
+// just does the math.
+func (me Prefix) Broadcast() Prefix {
+	network := me.addr.ui.or(me.Mask().ui.complement())
+	return Prefix{
+		addr: Address{
+			ui: network,
+		},
+		length: me.length,
+	}
+}
+
+// Host returns a new Prefix with the first `length` bits zeroed out so
+// that only the bits in the `host` part of the prefix are present
+func (me Prefix) Host() Prefix {
+	host := me.addr.ui.and(me.Mask().ui.complement())
+	return Prefix{
+		addr: Address{
+			ui: host,
+		},
+		length: me.length,
+	}
+}
+
+// String returns the string representation of this prefix in colon cidr
+// format (e.g 2001::1/64)
+func (me Prefix) String() string {
+	return fmt.Sprintf("%s/%d", me.addr.String(), me.Length())
+}
+
+// Uint64 returns the address and mask as uint64s
+func (me Prefix) Uint64() (addressHigh, addressLow, maskHigh, maskLow uint64) {
+	addressHigh, addressLow = me.addr.Uint64()
+	maskHigh, maskLow = me.Mask().Uint64()
+	return
+}
+
+// walkAddresses visits all of the addresses in the prefix in lexigraphical
+// order
+//
+// It returns false if iteration was stopped due to a callback return false or
+// true if it iterated all items.
+func (me Prefix) walkAddresses(callback func(Address) bool) bool {
```

Comment:
Walking Addresses in IPv6 *might* be a bad idea. What do you think?

Reply:
Yeah given how many addresses you could potentially be iterating over I don.t think its necessarily a good idea.